src/meta/client/MetaClient.cpp

```diff
@@ -957,5 +966,211 @@ folly::Future<StatusOr<bool>> MetaClient::heartbeat() {
+            value = *reinterpret_cast<const int64_t*>(item.get_value().data());
+            break;
+        case cpp2::ConfigType::BOOL:
+            eType = VariantTypeEnum::BOOL;
+            value = *reinterpret_cast<const bool*>(item.get_value().data());
+            break;
+        case cpp2::ConfigType::DOUBLE:
+            eType = VariantTypeEnum::DOUBLE;
+            value = *reinterpret_cast<const double*>(item.get_value().data());
+            break;
+        case cpp2::ConfigType::STRING:
+            eType = VariantTypeEnum::STRING;
+            value = item.get_value();
+            break;
+    }
+    return ConfigItem(item.get_space(), item.get_module(), item.get_name(), eType, value);
+}
+
+StatusOr<ConfigItem> MetaClient::getConfigFromCache(const std::string& space,
+                                                    const std::string& module,
+                                                    const std::string& name,
+                                                    const VariantTypeEnum& type) {
+    if (!configReady_) {
+        return Status::Error("Config cache not ready!");
+    }
+
+    auto ret = getSpaceIdByNameFromCache(space);
+    if (!ret.ok()) {
+        return ret.status();
+    }
+    {
+        folly::RWSpinLock::ReadHolder holder(configCacheLock_);
+        auto spaceIt = metaConfigMap_.find(space);
+        if (spaceIt == metaConfigMap_.end()) {
+            return Status::SpaceNotFound();
+        }
+        auto it = spaceIt->second.find({module, name});
+        if (it != spaceIt->second.end()) {
+            if (it->second.type_ != type) {
+                return Status::CfgErrorType();
+            }
+            return it->second;
+        }
+        return Status::CfgNotFound();
+    }
+}
+
+StatusOr<bool> MetaClient::setConfigToCache(const std::string& space, const std::string& module,
+                                            const std::string& name, const VariantType& value,
+                                            const VariantTypeEnum& type) {
+    if (!configReady_) {
+        return Status::Error("Config cache not ready!");
+    }
+
+    auto ret = getSpaceIdByNameFromCache(space);
+    if (!ret.ok()) {
+        return ret.status();
+    }
+    {
+        folly::RWSpinLock::WriteHolder holder(configCacheLock_);
+        std::pair<std::string, std::string> key = {module, name};
+        ConfigItem item(space, module, name, type, value);
+        metaConfigMap_[space].emplace(std::move(key), std::move(item));
+        return true;
+    }
+}
+
+StatusOr<std::vector<ConfigItem>>
+MetaClient::listConfigsFromCache(const std::string& space, const std::string& module) {
+    if (!configReady_) {
+        return Status::Error("Config cache not ready!");
+    }
+    std::vector<ConfigItem> ret;
+    {
+        folly::RWSpinLock::ReadHolder holder(configCacheLock_);
+        if (!space.empty()) {
+            for (const auto& entry : metaConfigMap_[space]) {
```

Comment:
why can use operator[] here?

Reply:
If we specify a space, only return configuration of this space. Whether read configs from cache or kvstore is still to be decided.